### PR TITLE
T4877: Added more checks if "import vrf" is used in bgp

### PR DESCRIPTION
**Pull request by aapostoliuk**
## Change Summary

1. Fixed: If rd and route-target are used in VRF, can not use "import vrf" in the same VRF in the same AFI/SAFI.
2. Fixed: If rd and route-target is used in VRF, this VRF can not be in the list of command "import vrf" in the same AFI/SAFI but in other VRFs.
3. Fixed: Do not allow to delete vrf if it is used in import list of other vrfs.
4. Added smoketests to check "import vrf" issues.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes)
- [ ] Migration from an old Vyatta component to vyos-1x, please link to related PR inside obsoleted component
- [ ] Other (please describe):

## Related Task(s)

* https://phabricator.vyos.net/T4877

## Component(s) name

bgp

## Proposed changes

1. Fixed: If rd and route-target are used in VRF, can not use "import vrf" in the same VRF in the same AFI/SAFI.
2. Fixed: If rd and route-target is used in VRF, this VRF can not be in the list of command "import vrf" in the same AFI/SAFI but in other VRFs.
3. Fixed: Do not allow to delete vrf if it is used in import list of other vrfs.
4. Added smoketests to check "import vrf" issues.

## How to test

1. Fixed: If rd and route-target are used in VRF, can not use "import vrf" in the same VRF in the same AFI/SAFI.
Configuration:
```
set vrf name TEST1 protocols bgp system-as '65011'
set vrf name TEST1 table '222'
set vrf name TEST2 protocols bgp address-family ipv4-unicast import vrf 'TEST1'
set vrf name TEST2 protocols bgp system-as '65012'
set vrf name TEST2 table '223'
set vrf name TEST2 protocols bgp address-family ipv4-unicast rd vpn export 65012:20
commit
```

Before:
```
vyos@vyos# commit
[ vrf name TEST2 protocols bgp ]
VyOS had an issue completing a command.

We are sorry that you encountered a problem while using VyOS.
There are a few things you can do to help us (and yourself):
- Contact us using the online help desk if you have a subscription:
  https://support.vyos.io/
- Make sure you are running the latest version of VyOS available at:
  https://vyos.net/get/
- Consult the community forum to see how to handle this issue:
  https://forum.vyos.io
- Join us on Slack where our users exchange help and advice:
  https://vyos.slack.com

When reporting problems, please include as much information as possible:
- do not obfuscate any data (feel free to contact us privately if your
  business policy requires it)
- and include all the information presented below

Report time:      2022-12-29 14:09:38
Image version:    VyOS 1.4-rolling-202212060318
Release train:    current

Built by:         autobuild@vyos.net
Built on:         Tue 06 Dec 2022 03:18 UTC
Build UUID:       c04b006d-fd52-4651-8c48-317ebc9a286e
Build commit ID:  0f1eaa32bd8829

Architecture:     x86_64
Boot via:         installed image
System type:      KVM guest

Hardware vendor:  QEMU
Hardware model:   Standard PC (i440FX + PIIX, 1996)
Hardware S/N:
Hardware UUID:    4765dc56-c405-4447-ad9d-6c8f91c21725

Traceback (most recent call last):
  File "/usr/libexec/vyos/conf_mode/protocols_bgp.py", line 388, in <module>
    apply(c)
  File "/usr/libexec/vyos/conf_mode/protocols_bgp.py", line 379, in apply
    frr_cfg.commit_configuration(bgp_daemon)
  File "/usr/lib/python3/dist-packages/vyos/frr.py", line 480, in commit_configuration
    raise ConfigurationNotValid(f'Config commit retry counter ({count_max}) exceeded for {daemon} dameon!')
vyos.frr.ConfigurationNotValid: Config commit retry counter (5) exceeded for bgpd dameon!



[[vrf name TEST2 protocols bgp]] failed
Commit failed
[edit]

```
After

```
Please unconfigure "import vrf" commands before using vpn commands in
the same VRF!
```

2. Fixed: If rd and route-target is used in VRF, this VRF can not be in the list of command "import vrf" in the same AFI/SAFI but in other VRFs.
Configuration:
```
set vrf name TEST1 protocols bgp system-as '65011'
set vrf name TEST1 table '222'
set vrf name TEST2 protocols bgp address-family ipv4-unicast import vrf 'TEST1'
set vrf name TEST2 protocols bgp system-as '65012'
set vrf name TEST2 table '223'
set vrf name TEST1 protocols bgp address-family ipv4-unicast rd vpn export 65011:20
```

Before
```
vyos@vyos# commit
[ vrf name TEST1 protocols bgp ]
VyOS had an issue completing a command.

We are sorry that you encountered a problem while using VyOS.
There are a few things you can do to help us (and yourself):
- Contact us using the online help desk if you have a subscription:
  https://support.vyos.io/
- Make sure you are running the latest version of VyOS available at:
  https://vyos.net/get/
- Consult the community forum to see how to handle this issue:
  https://forum.vyos.io
- Join us on Slack where our users exchange help and advice:
  https://vyos.slack.com

When reporting problems, please include as much information as possible:
- do not obfuscate any data (feel free to contact us privately if your
  business policy requires it)
- and include all the information presented below

Report time:      2022-12-29 14:14:17
Image version:    VyOS 1.4-rolling-202212060318
Release train:    current

Built by:         autobuild@vyos.net
Built on:         Tue 06 Dec 2022 03:18 UTC
Build UUID:       c04b006d-fd52-4651-8c48-317ebc9a286e
Build commit ID:  0f1eaa32bd8829

Architecture:     x86_64
Boot via:         installed image
System type:      KVM guest

Hardware vendor:  QEMU
Hardware model:   Standard PC (i440FX + PIIX, 1996)
Hardware S/N:
Hardware UUID:    4765dc56-c405-4447-ad9d-6c8f91c21725

Traceback (most recent call last):
  File "/usr/libexec/vyos/conf_mode/protocols_bgp.py", line 388, in <module>
    apply(c)
  File "/usr/libexec/vyos/conf_mode/protocols_bgp.py", line 379, in apply
    frr_cfg.commit_configuration(bgp_daemon)
  File "/usr/lib/python3/dist-packages/vyos/frr.py", line 480, in commit_configuration
    raise ConfigurationNotValid(f'Config commit retry counter ({count_max}) exceeded for {daemon} dameon!')
vyos.frr.ConfigurationNotValid: Config commit retry counter (5) exceeded for bgpd dameon!



[[vrf name TEST1 protocols bgp]] failed
Commit failed
[edit]

```
After 
```
vyos@vyos# commit

Command "import vrf" conflicts with "rd vpn export" command!

[[vrf name TEST1 protocols bgp]] failed
Commit failed
[edit]

```

3. Fixed: Do not allow to delete vrf if it is used in import list of other vrfs.

Configuration
```
set vrf name TEST1 protocols bgp system-as '65011'
set vrf name TEST1 table '222'
set vrf name TEST2 protocols bgp address-family ipv4-unicast import vrf 'TEST1'
set vrf name TEST2 protocols bgp system-as '65012'
set vrf name TEST2 table '223'
commit
```
before
`vyos@vyos# delete vrf name TEST1`

commit passed

After:
```
vyos@vyos# commit

Cannot delete vrf TEST1 instance, Please unconfigure import vrf
commands!

delete [ vrf name TEST1 protocols bgp ] failed
[[vrf]] failed
Commit failed
[edit]
```

## Checklist:

- [x] I have read the [**CONTRIBUTING**](https://github.com/vyos/vyos-1x/blob/current/CONTRIBUTING.md) document
- [x] I have linked this PR to one or more Phabricator Task(s)
- [x] I have run the components [**SMOKETESTS**](https://github.com/vyos/vyos-1x/tree/current/smoketest/scripts/cli) if applicable
- [x] My commit headlines contain a valid Task id
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

### CODE REVIEW

**Comment by c-po** [comment]
Thanks for the contribution. Inside the smoketests you only validate if the CLI config applies or not. You should also verify if the applied configuration really shows up in bgpd. See all the other testcases when we assert if certain options are present in vtysh or not.

**Comment by aapostoliuk** [comment]
> Thanks for the contribution. Inside the smoketests you only validate if the CLI config applies or not. You should also verify if the applied configuration really shows up in bgpd. See all the other testcases when we assert if certain options are present in vtysh or not.

**Comment by aapostoliuk** [comment]
> Thanks for the contribution. Inside the smoketests you only validate if the CLI config applies or not. You should also verify if the applied configuration really shows up in bgpd. See all the other testcases when we assert if certain options are present in vtysh or not.

@c-po 
All commands that I used are tested in tests 10 and 13
https://github.com/vyos/vyos-1x/blob/current/smoketest/scripts/cli/test_protocols_bgp.py#L767
https://github.com/vyos/vyos-1x/blob/current/smoketest/scripts/cli/test_protocols_bgp.py#L857
Or do you mean I need to verify the applied config after a successful commit?
as below:
https://github.com/vyos/vyos-1x/pull/1728/files#diff-b7b829f1104c89e4b23384faad9b0d18c3318c76c32ac495e0234a6b1b06c2b4R995

**Comment by aapostoliuk** [comment]
The purpose of new tests is to check CLI options compatibility only, therefore actual FRR config is out of the scope here.

**Comment by c-po** [comment]
> The purpose of new tests is to check CLI options compatibility only, therefore actual FRR config is out of the scope here.

But checking if your option before generating the exception was properly added to the FRR configuration should be part of the test, too. It does not harm - it only makes VyOS better.

**Comment by aapostoliuk** [comment]
> > The purpose of new tests is to check CLI options compatibility only, therefore actual FRR config is out of the scope here.
> 
> But checking if your option before generating the exception was properly added to the FRR configuration should be part of the test, too. It does not harm - it only makes VyOS better.

Done

File: smoketest/scripts/cli/test_protocols_bgp.py [hunk 1]
```diff
@@ -13,7 +13,7 @@
 #
 # You should have received a copy of the GNU General Public License
 # along with this program.  If not, see <http://www.gnu.org/licenses/>.
-
+import pprint
```
Comment (sever-sever) [inline smoketest/scripts/cli/test_protocols_bgp.py]: Why do we need it?

**Review by sever-sever — CHANGES_REQUESTED** [review]

File: smoketest/scripts/cli/test_protocols_bgp.py (hunk 1 above)
Comment (aapostoliuk) [reply smoketest/scripts/cli/test_protocols_bgp.py]: fixed

**Review by c-po — APPROVED** [review]

**Pull request merged**